csharp/ql/lib/semmle/code/csharp/exprs/Call.qll

```diff
@@ -290,7 +303,54 @@ private predicate isValidExplicitParamsType(Parameter p, Type t) {
  * ```
  */
 class MethodCall extends Call, QualifiableExpr, LateBindableExpr, @method_invocation_expr {
-  override Method getTarget() { expr_call(this, result) }
+  override Method getTarget() {
+    expr_call(this, result)
+    or
+    result = this.getACandidateTarget()
+  }
+
+  private Method getACandidateTarget() {
+    not exists(Method m | expr_call(this, m)) and
+    exists(string name, ValueOrRefType alternativeQualifierType |
+      invocation_member_name(this, name) and
+      alternativeQualifierType =
+        this.getQualifier()
+            .getType()
+            .(ValueOrRefType)
+            .getABaseType*()
+            .getAnAmbiguousAlternativeType*() and
+      result = alternativeQualifierType.getAMethod(name) and
+      // todo: enable the same logic for extension methods too.
+      not result.isExtensionMethod() and
+      // Make sure all parameters have a corresponding argument.
+      forall(Parameter p | result.getAParameter() = p |
+        exists(Expr arg | arg = this.getAnArgument() |
+          this.isParameterArgumentAssignable(result, p, arg)
+        )
+        or
+        p.hasDefaultValue() // call to `M(int p = 1)` might not have any arguments
+        or
+        p.isParams() // call to `M(params int[] p)` might not have any arguments
+      )
+    ) and
+    // Make sure all arguments have a corresponding parameter.
+    forall(Expr arg | this.getAnArgument() = arg |
+      exists(Parameter p | result.getAParameter() = p |
+        this.isParameterArgumentAssignable(result, p, arg)
+      )
+    )
+  }
+
+  private predicate isParameterArgumentAssignable(Callable c, Parameter p, Expr arg) {
+    arg = this.getArgumentForParameterInternal(c, p) and
+    (
+      not p.isParams() and
+      arg.getType().isImplicitlyConvertibleTo(p.getType())
+      or
+      p.isParams() and
+      arg.getType().isImplicitlyConvertibleTo(p.getType().(ArrayType).getElementType())
```

Comment:
I don't think this works if either the parameter type or the argument type contains type parameters.

Reply:
We can definitely improve this further. At the same time, I don't think it's critical. It means that there are cases when we don't find a candidate call target, doesn't it? The opposite can also happen: we may find too many candidate call targets. These are expected as we haven't implemented the entire resolution logic.